### PR TITLE
Use %triggerun to enable migration flag for all pre-1.8.0 upgrades

### LINKED ISSUE #1780: Enable migration flag when upgrading from all pre-1.8.0 releases
Because of the trixie update (see https://github.com/freedomofpress/securedrop-workstation/issues/1763), we need to ensure that when upgrading from any pre-1.8.0 release, we run the full migration.

### PR DESCRIPTION
**Pull request by legoktm**
_This is rebased on top of #1784, which must be merged first_

We want all upgrades from a pre-1.8.0 version to trigger a full migration for the trixie switch. %triggerun is a good fit for this as per rpm-scriptlets(7), which says triggers are "often used for version specific migration tasks".

Fixes #1780.

## Test plan

Pre-1.8.0 case:
* [x] Manually fetch 1.7.1 RPM and install it (running --apply is not necessary)
* [x] Build this PR via `make clone` and manually install the RPM with `sudo dnf install ./path....`
* [x] Run `sudo qubes-dom0-update` and install the new RPM; observe that migration flag is created
* [ ] (added by @deeplow) `sudo qubesctl top.enabled` should NOT list `/srv/salt/_tops/base/securedrop_salt.sd-workstation.top`

Post-1.8.0 case:
* [x] Install 1.8.0 RPM (running --apply is not necessary)
* [x] Build this PR via `make clone` and manually install the RPM with `sudo dnf install ./path....`
* [ ] Run `sudo qubes-dom0-update` and install the new RPM; observe that migration flag is NOT created
* [ ] (added by @deeplow) `sudo qubesctl top.enabled` should list `/srv/salt/_tops/base/securedrop_salt.sd-workstation.top`

## Checklist

This change accounts for:
- [ ] any necessary RPM packaging updates (e.g., added/removed files, see `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`)
- [ ] any required documentation

### CODE REVIEW

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec [hunk 1]
```diff
@@ -220,6 +215,12 @@ if [ $1 -eq 0 ]; then
     %systemd_user_preun sdw-notify.timer
 fi
 
+# Force a full migration run for all pre-1.8.0 upgrades
+%triggerun -- %{name} < 1.8.0
+mkdir -p /tmp/sdw-migrations
+touch /tmp/sdw-migrations/debian-13-bump
+qubesctl top.disable securedrop_salt.sd-workstation > /dev/null ||:
```
[the remark this answers is not in the document]
Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec]: No, these need to be hard failures, we can't silently skip the migration.

[the remark this answers is not in the document]
Comment (deeplow) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec]: Maybe even this last one could fail. With that said, I haven't tested what happens if it blows up here.

```suggestion
%triggerun -- %{name} < 1.8.0
mkdir -p /tmp/sdw-migrations
touch /tmp/sdw-migrations/debian-13-bump
qubesctl top.disable securedrop_salt.sd-workstation > /dev/null
```

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec [hunk 2]
```diff
@@ -191,11 +191,6 @@ install -m 644 securedrop_salt/apt-test_freedom_press.sources.j2 %{buildroot}/sr
 # Update Salt Configuration
 qubesctl saltutil.clear_cache -l quiet --out quiet > /dev/null || true
 qubesctl saltutil.sync_all refresh=true -l quiet --out quiet > /dev/null || true
-qubesctl top.disable securedrop_salt.sd-workstation > /dev/null ||:
```
Comment (deeplow) [inline rpm-build/SPECS/securedrop-workstation-dom0-config.spec:194]: We need to re-enable this, as otherwise non-migration releases will not run any dom0 highstate at all.

Maybe we should revert https://github.com/freedomofpress/securedrop-workstation/pull/1772/ (which also removes it from `sdw-admin --apply`)

**Review by deeplow — CHANGES_REQUESTED** [review]

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 2 above)
Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:194]: So should we have a %triggerrun >= 1.8.0 that does that? I think we still need to keep the disable for anyone on an old updater, right?

Comment (deeplow) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:194]: Oh. Right. We do need to make it that way.

Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:194]: Actually I'm not sure why we need the top.enable. Here's the flow:

1. Normally top file is enabled
2. On an upgrade from 1.7.1 -> 1.8.x, %post/%trigger disables the top. Updater invokes sdw-admin --apply which re-enables it.
3. Top file is enabled once again
4. Future update skips the %triggerun, so it stays enabled. (plus sdw-admin --apply will always re-enable it anyways)

So why do we need an explicit enable for other releases?

I think the main thing I realized, which I'll do now, is that the top.disable should be its own %triggerun, which is always < 1.8.0, while the migration flag will be moving a moving target.

Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:194]: Actually no, we can only have one %triggerun. I think we're okay for now but let me know if I'm overlooking something.

Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:194]: Per discussion, I've added a step to re-enable the to file into the updater. I rebased it on top of #1784 just because it's touching the same area of code.

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec [hunk 3]
```diff
@@ -191,11 +191,6 @@ install -m 644 securedrop_salt/apt-test_freedom_press.sources.j2 %{buildroot}/sr
...
-
-# Force full run of all Salt states - uncomment in release branch
-mkdir -p /tmp/sdw-migrations
-touch /tmp/sdw-migrations/debian-13-bump
```
Comment (deeplow) [inline rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: I think we should leave this here, but commented out. as we had before. Better to recognize the "toggle" than have to recall how to enable the flag and where to put it.

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 1 above)
[the remark this answers is not in the document]
Comment (deeplow) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec]: Not sure if you saw my comment here bellow the AI's before you closed it. In case you did, is the concern that it's better to proceed even the top failed to be disabled? It really shouldn't run into this case, and not such a big deal. If it doesn't fail here, it'll fail later down the line, but failing here may be easier to diagnose than having to realize that the top had failed to be disabled.

File: launcher/tests/test_updaterapp.py [hunk 4]
```diff
@@ -30,6 +30,7 @@ def test_run_full_update_dom0_update_failure_exits_early(
     assert not apply_dom0_state_mock.called
 
 
+@mock.patch("sdw_updater.Updater.enable_dom0_state", return_value=UpdateStatus.UPDATES_OK)
```
Comment (deeplow) [inline launcher/tests/test_updaterapp.py:33]: It could make sense to have a test (or add in an existing one) some check that enable the `sd-workstation.top` before any other salt operation. Basically beofre the migration step and/or dom0 highstate. Otherwise we could have a situation where tests still pass because it's in fact enabled, but in reality it fails as it doesn't happen at the right point in time.

**Review by deeplow — COMMENTED** [review]
Minor changes only, in order to catch some edge-cases and development ergonomics.

Also, I felt the need to modify the test plan since the last changes. E.g. now it includes updater changes and it seemed worth evaluating the actual previous updater behavior in these cases, as it already has quite a few possible paths.

## Modified Test plans


@legoktm please do give this test plan a pass just in case there was an obvious thing I missed.

### Pre-1.8.0 case:
> 
>     * [ ]  Manually fetch 1.7.1 RPM and install it (running --apply is not necessary)

`--apply` ended up being necessary now that an updater change was included. Otherwise it would fail because the VMs it's expecting don't exist (e.g. the old templates)

>     * [ ]  Build this PR via `make clone` and manually install the RPM with `sudo dnf install ./path....`

I did make clone, and then served via the `./scripts/local-yum-server` (I had to rebase this on main for local testing to grab that file)

>     * [ ]  Run `sudo qubes-dom0-update` and install the new RPM; observe that migration flag is created

I ran instead `sdw-updater --skip-delta 0`.

>     * [ ]  (added by @deeplow) `sudo qubesctl top.enabled` should NOT list `/srv/salt/_tops/base/securedrop_salt.sd-workstation.top`

- `top.enabled`: enabled -> disabled -> enabled
   - :information_source: this must have been the code in `sdw-admin --apply` because at this stage the updater hasn't run yet.
- migration flag: `absent -> present -> absent

### Post-1.8.0 case:

>     * [ ]  Install 1.8.0 RPM (running --apply is not necessary)

same as in the other case. I ran `sdw-admin --apply`. Then I removed the flag (otherwise I'd need to reboot)

>     * [ ]  Build this PR via `make clone` and manually install the RPM with `sudo dnf install ./path....`

(Same as the other case) I did make clone, and then served via the `./scripts/local-yum-server` (I had to rebase this on main for local testing to grab that file).

>     * [ ]  Run `sudo qubes-dom0-update` and install the new RPM; observe that migration flag is NOT created

(same as other test case) I ran instead `sdw-updater --skip-delta 0`.

>     * [ ]  (added by @deeplow) `sudo qubesctl top.enabled` should list `/srv/salt/_tops/base/securedrop_salt.sd-workstation.top`

- [x] migration flag was never set throughout the process
- [x] `sd-workstation.top` never got disabled
- [x] Updater only applies dom0 state

  <details>
  
  ```
  [user@dom0 ~]$ tail -f ~/.securedrop_updater/logs/updater.log 
  2026-07-27 15:36:52,173 - sd:52(main) INFO: Starting SecureDrop Launcher
  2026-07-27 15:36:52,174 - sd.sdw_updater.Updater:378(read_dom0_update_flag_from_disk) INFO: Cannot read dom0 status flag, assuming first run
  2026-07-27 15:36:52,174 - sd.sdw_updater.Updater:493(should_launch_updater) INFO: Update status not available, launching updater.
  2026-07-27 15:37:08,360 - sd.sdw_updater.UpdaterApp:140(_check_network_and_update) INFO: Network check successful; checking for updates.
  2026-07-27 15:37:08,360 - sd.sdw_updater.UpdaterApp:160(apply_all_updates) INFO: Starting UpgradeThread
  2026-07-27 15:37:08,362 - sd.sdw_updater.Updater:116(apply_updates_dom0) INFO: Applying all updates to dom0
  2026-07-27 15:37:18,142 - sd.sdw_updater.Updater:260(_check_updates_dom0) ERROR: dom0 updates required, or cannot check for updates
  2026-07-27 15:37:18,143 - sd.sdw_updater.Updater:261(_check_updates_dom0) ERROR: Command '['sudo', 'qubes-dom0-update', '--check-only']' returned non-zero exit status 100.
  2026-07-27 15:37:18,143 - sd.sdw_updater.Updater:273(_apply_updates_dom0) INFO: Updating dom0
  2026-07-27 15:37:43,304 - sd.sdw_updater.Updater:281(_apply_updates_dom0) INFO: dom0 updates have been applied and a reboot is required.
  2026-07-27 15:37:43,305 - sd.sdw_updater.Updater:425(apply_dom0_state) INFO: Applying dom0 state
  2026-07-27 15:38:08,597 - sd.sdw_updater.Updater:430(apply_dom0_state) INFO: Dom0 state applied
  2026-07-27 15:38:08,598 - sd.sdw_updater.Updater:133(apply_updates_templates) INFO: Applying all updates to VMs: sd-viewer-debian-13, sd-inbox-debian-13, fedora-43-xfce
  2026-07-27 15:38:08,978 - sd.sdw_updater.Updater:230(_qubes_updater_parse_progress) INFO: Starting update on template: 'sd-viewer-debian-13'
  2026-07-27 15:38:08,979 - sd.sdw_updater.Updater:230(_qubes_updater_parse_progress) INFO: Starting update on template: 'fedora-43-xfce'
  2026-07-27 15:38:08,979 - sd.sdw_updater.Updater:230(_qubes_updater_parse_progress) INFO: Starting update on template: 'sd-inbox-debian-13'
  2026-07-27 15:38:28,603 - sd.sdw_updater.Updater:242(_qubes_updater_parse_progress) INFO: Update successful for template: 'sd-viewer-debian-13'
  2026-07-27 15:38:30,579 - sd.sdw_updater.Updater:242(_qubes_updater_parse_progress) INFO: Update successful for template: 'sd-inbox-debian-13'
  2026-07-27 15:38:34,799 - sd.sdw_updater.Updater:242(_qubes_updater_parse_progress) INFO: Update successful for template: 'fedora-43-xfce'
  2026-07-27 15:38:35,617 - sd.sdw_updater.Updater:378(read_dom0_update_flag_from_disk) INFO: Cannot read dom0 status flag, assuming first run
  2026-07-27 15:38:35,617 - sd.sdw_updater.Updater:156(apply_updates_templates) INFO: Template updates successful
  2026-07-27 15:38:35,617 - sd.sdw_updater.Updater:378(read_dom0_update_flag_from_disk) INFO: Cannot read dom0 status flag, assuming first run
  2026-07-27 15:38:35,617 - sd.sdw_updater.Updater:311(_write_updates_status_flag_to_disk) INFO: Setting update flag to 2 in dom0
  2026-07-27 15:38:35,618 - sd.sdw_updater.Updater:294(_write_last_updated_flags_to_disk) INFO: Setting last updated to 2026-07-27 15:38:35 in dom0
  2026-07-27 15:38:35,618 - sd.sdw_updater.UpdaterApp:82(upgrade_status) INFO: Signal: upgrade_status {'dom0': <UpdateStatus.REBOOT_REQUIRED: '2'>, 'apply_dom0': <UpdateStatus.UPDATES_OK: '0'>, 'apply_all': <UpdateStatus.UPDATES_OK: '0'>, 'templates': <UpdateStatus.UPDATES_OK: '0'>, 'recommended_action': <UpdateStatus.REBOOT_REQUIRED: '2'>}
  2026-07-27 15:38:35,624 - sd.sdw_updater.UpdaterApp:87(upgrade_status) INFO: Reboot required
  ```

  </details> 


---

## Extra test cases:
### Post 1.8.0 -> post 1.8.0 But with migration

This extra case was needed to:
- test post-1.8.0 migration path
- test updater changes introduced (positive path, where it doesn't make that much of a difference)

Test instructions:
- [x] have RPM this branch installed (could be via the updater already)
- [x] In `sd-dev` :
  - [x] simulate a new update w/ migration:`./update-version.py 1.10.0`
  - [x] Add migration
    <details><summary>This is what my patch looked like</summary>
    
    ```diff
    diff --git a/VERSION b/VERSION
    index 6ed8c32..81c871d 100644
    --- a/VERSION
    +++ b/VERSION
    @@ -1 +1 @@
    -1.9.0rc1
    +1.10.0
    diff --git a/rpm-build/SPECS/securedrop-workstation-dom0-config.spec b/rpm-build/SPECS/securedrop-workstation-dom0-config.spec
    index d3b2fe4..ee2ddeb 100644
    --- a/rpm-build/SPECS/securedrop-workstation-dom0-config.spec
    +++ b/rpm-build/SPECS/securedrop-workstation-dom0-config.spec
    @@ -1,5 +1,5 @@
     Name:          securedrop-workstation-dom0-config
    -Version:       1.9.0rc1
    +Version:       1.10.0
     Release:       1%{?dist}
     Summary:       SecureDrop Workstation
     
    @@ -192,12 +192,16 @@ install -m 644 securedrop_salt/apt-test_freedom_press.sources.j2 %{buildroot}/sr
     qubesctl saltutil.clear_cache -l quiet --out quiet > /dev/null || true
     qubesctl saltutil.sync_all refresh=true -l quiet --out quiet > /dev/null || true
     
    +mkdir -p /tmp/sdw-migrations
    +touch /tmp/sdw-migrations/post-1.8.0-migration
    +
     # Enable service that conditionally removes our systemd-logind customizations
     # on dev machines only.
     # It's clumsy, but overrides to systemd services can't be conditionally applied.
     # Changes take place after systemd restart.
     systemctl enable securedrop-logind-override-disable.service ||:
     
    +
     # Customize xfce power settings and icon size.
     # Power settings changes conditionally disabled in dev environments.
     %systemd_user_post securedrop-user-xfce-icon-size.service
    @@ -223,6 +227,9 @@ touch /tmp/sdw-migrations/debian-13-bump
     qubesctl top.disable securedrop_salt.sd-workstation > /dev/null ||:
     
     %changelog
    +* Mon Jul 27 2026 SecureDrop Team <securedrop@freedom.press> - 1.10.0
    +- See changelog.md
    +
     * Mon Jul 20 2026 SecureDrop Team <securedrop@freedom.press> - 1.9.0rc1
     - See changelog.md
 

    ```
    </details>
- [ ] In `dom0`:
  - [ ] `make clone && ./scripts/local-rpm-server.sh`
  - [ ] `sdw-updater --skip-delta 0`
  - [ ] updater logs show: migration detected and applied
  - [ ] updater logs show: `dom0 top file enabled`


---
### Post 1.8.0 -> post 1.8.0 (no migration)

Set up:
- [x] Have a post-1.8.0 version already installed
- [x] Ensure migration flags have been properly cleaned (otherwise `sudo rm -r /tmp/sdw-migrations/`
- [x] In `dom0`:
  - [x] `make clone && ./scripts/local-rpm-server.sh`
- [x] Explicitly disable top file: `sudo qubesctl top.disable securedrop_salt.sd-workstation` (simulate some failure to re-enable this)

Testing:
- [x] `sdw-updater --skip-delta 0`
- [x] confirm `sd-workstation.top` enabled by the end of the updater run: check in `sudo qubesctl top.enabled`
- [x] confirm dom0 highstate ran (e.g  `updater-detail.log`)

<details><summary>This is what my updater log looked like</summary>

```
2026-07-27 16:27:57,507 - sd:52(main) INFO: Starting SecureDrop Launcher
2026-07-27 16:27:57,510 - sd.sdw_updater.Updater:493(should_launch_updater) INFO: Update interval expired: launching updater.
2026-07-27 16:28:02,033 - sd.sdw_updater.UpdaterApp:140(_check_network_and_update) INFO: Network check successful; checking for updates.
2026-07-27 16:28:02,034 - sd.sdw_updater.UpdaterApp:160(apply_all_updates) INFO: Starting UpgradeThread
2026-07-27 16:28:02,035 - sd.sdw_updater.Updater:116(apply_updates_dom0) INFO: Applying all updates to dom0
2026-07-27 16:28:09,471 - sd.sdw_updater.Updater:260(_check_updates_dom0) ERROR: dom0 updates required, or cannot check for updates
2026-07-27 16:28:09,471 - sd.sdw_updater.Updater:261(_check_updates_dom0) ERROR: Command '['sudo', 'qubes-dom0-update', '--check-only']' returned non-zero exit status 100.
2026-07-27 16:28:09,471 - sd.sdw_updater.Updater:273(_apply_updates_dom0) INFO: Updating dom0
2026-07-27 16:28:34,593 - sd.sdw_updater.Updater:281(_apply_updates_dom0) INFO: dom0 updates have been applied and a reboot is required.
2026-07-27 16:28:34,594 - sd.sdw_updater.Updater:424(enable_dom0_state) INFO: Enabling dom0 top file
2026-07-27 16:28:37,632 - sd.sdw_updater.Updater:429(enable_dom0_state) INFO: dom0 top file enabled
2026-07-27 16:28:37,633 - sd.sdw_updater.Updater:448(apply_dom0_state) INFO: Applying dom0 state
2026-07-27 16:29:03,302 - sd.sdw_updater.Updater:453(apply_dom0_state) INFO: Dom0 state applied
2026-07-27 16:29:03,303 - sd.sdw_updater.Updater:133(apply_updates_templates) INFO: Applying all updates to VMs: sd-viewer-debian-13, fedora-43-xfce, sd-inbox-debian-13
2026-07-27 16:29:03,760 - sd.sdw_updater.Updater:230(_qubes_updater_parse_progress) INFO: Starting update on template: 'fedora-43-xfce'
2026-07-27 16:29:03,761 - sd.sdw_updater.Updater:230(_qubes_updater_parse_progress) INFO: Starting update on template: 'sd-inbox-debian-13'
2026-07-27 16:29:03,762 - sd.sdw_updater.Updater:230(_qubes_updater_parse_progress) INFO: Starting update on template: 'sd-viewer-debian-13'
2026-07-27 16:29:23,894 - sd.sdw_updater.Updater:242(_qubes_updater_parse_progress) INFO: Update successful for template: 'sd-viewer-debian-13'
2026-07-27 16:29:26,962 - sd.sdw_updater.Updater:242(_qubes_updater_parse_progress) INFO: Update successful for template: 'sd-inbox-debian-13'
2026-07-27 16:29:30,774 - sd.sdw_updater.Updater:242(_qubes_updater_parse_progress) INFO: Update successful for template: 'fedora-43-xfce'
2026-07-27 16:29:31,312 - sd.sdw_updater.Updater:158(apply_updates_templates) INFO: Template updates failed
2026-07-27 16:29:31,313 - sd.sdw_updater.Updater:311(_write_updates_status_flag_to_disk) INFO: Setting update flag to 2 in dom0
2026-07-27 16:29:31,313 - sd.sdw_updater.Updater:294(_write_last_updated_flags_to_disk) INFO: Setting last updated to 2026-07-27 16:29:31 in dom0
2026-07-27 16:29:31,314 - sd.sdw_updater.UpdaterApp:82(upgrade_status) INFO: Signal: upgrade_status {'dom0': <UpdateStatus.REBOOT_REQUIRED: '2'>, 'enable_dom0': <UpdateStatus.UPDATES_OK: '0'>, 'apply_dom0': <UpdateStatus.UPDATES_OK: '0'>, 'apply_all': <UpdateStatus.UPDATES_OK: '0'>, 'templates': <UpdateStatus.REBOOT_REQUIRED: '2'>, 'recommended_action': <UpdateStatus.REBOOT_REQUIRED: '2'>}
2026-07-27 16:29:31,315 - sd.sdw_updater.UpdaterApp:87(upgrade_status) INFO: Reboot required

```

</details>

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 1 above)
[the remark this answers is not in the document]
Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec]: Sorry I missed it. I agree, we should let it fail here. I also think we should get rid of the > /dev/null to make it a bit easier to figure out when something actually does go wrong.

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 3 above)
Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: I don't think we need it; my plan is that we only ever use the %triggerun now, and the only thing that needs doing is adjusting the version when it runs.

File: launcher/tests/test_updaterapp.py (hunk 4 above)
Comment (legoktm) [reply launcher/tests/test_updaterapp.py:33]: Done, added a test that verifies the call order looks like:
```
    assert manager.mock_calls == [
        mock.call.enable_dom0_state(),
        mock.call.migration_is_required(),
        mock.call.apply_dom0_state(),
    ]

```

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 3 above)
Comment (deeplow) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: OK. I was not reviewing with this expectation in mind, then. Could you give an example how it would look for when we do a future release with migration vs no migration?

I may be misunderstanding your goal here, but I see a migration more to do with the needs of the current release rather than which release users come from.

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 1 above)
[the remark this answers is not in the document]
Comment (deeplow) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec]: Thanks. Error messages should still show with `>/dev/null` because it's stderr. So I think it's fine leave it as well.  But without it is fine also as `top.enabled` (unlike everything else salt-related) is not too verbose.

File: launcher/tests/test_updaterapp.py (hunk 4 above)
Comment (deeplow) [reply launcher/tests/test_updaterapp.py:33]: Thanks! That's exactly what I was looking for.

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 3 above)
Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: The idea is that if you are coming from any release before 1.8.0, you need a full migration run. 

Consider the following releases:
* 1.7.0: uses bookworm templates
* 1.7.1: bugfix release
* 1.8.0: switches to trixie, needs full migration
* 1.8.1: bugfix release

Anyone upgrading from < 1.8.0 to either 1.8.0 or 1.8.1 needs a full migration run. But if you go from 1.8.0 -> 1.8.1, you don't need it. The way the triggerun is set up, it handles both the 1.7.1 -> 1.8.1 and 1.8.0 -> 1.8.1 case optimally, only the former needs the full migration and gets it.

(The previous approach was broken for "skip" upgrades, e.g. if we only set the migration flag in 1.8.0 and not 1.8.1, then someone upgrading 1.7.1 -> 1.8.1 would never get the migration flag and the upgrade would break.) 

Now lets say in the future we release:
* 1.9.0: Fedora 44, needs full migration

At this point we edit the %triggerun condition to be < 1.9.0, because everyone before that needs the full migration, whether it's 1.7.1 -> 1.9.0 or 1.8.1 -> 1.9.0.

So it's still tied to whether a specific release needs a migration, but it doesn't become a thing we toggle per-release anymore. Does that make sense?

Comment (deeplow) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: OK. Then I get the point now. Also, could we have an issue for this? It feels like it should have bigger visibility as we're not just fixing the "pre-1.8.0 upgrades" here.

And I'd ask if we could add some instructional comments explaining how migrations are triggered now / when to bump the version number and to what. Because from the comment there it's not very obvious what one should do.

Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: Filed https://github.com/freedomofpress/securedrop-workstation/issues/1815.

Sure, I can improve the comments.

Comment (deeplow) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: Thanks. Just ping me when done and I'll approve it!

Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: @deeplow done

**Comment by legoktm** [comment]
I tried restarting just the securedrop_test_dom0 job via the OpenQA website a few times, but it keeps failing so I'm going to retrigger the entire sequence via GH.

**Comment by deeplow** [comment]
I see [this run](https://openqa.qubes-os.org/tests/191243/logfile?filename=serial_terminal.txt#line-166) failed with:

```
Failed to download metadata (baseurl: "https://yum-test.securedrop.org/workstation/dom0/r4.3-nightlies") for repository "securedrop-workstation-dom0": Usable URL not found
_[?25hmake: *** [Makefile:158: install-dom0-test-prereqs] Error 1
```

Could it be that somehow this worker has some problem reaching cloudflare? I'll restart

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec [hunk 5]
```diff
@@ -220,6 +215,16 @@ if [ $1 -eq 0 ]; then
     %systemd_user_preun sdw-notify.timer
 fi
 
+# This trigger allows us to conditionally run code based on the
+# version we are upgrading from. The version conditional should
+# be bumped whenever there is some change in the release that
+requires a full `sdw-admin --apply` run.
```
Comment (deeplow) [inline rpm-build/SPECS/securedrop-workstation-dom0-config.spec]:
```suggestion
# requires a full `sdw-admin --apply` run.
```

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 3 above)
Comment (deeplow) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec:198]: Thanks. Minor typo there (missing `#`)

File: rpm-build/SPECS/securedrop-workstation-dom0-config.spec (hunk 5 above)
Comment (legoktm) [reply rpm-build/SPECS/securedrop-workstation-dom0-config.spec]: omg whoops. I wonder why the rpm spec linter didn't catch this

**Review by deeplow — APPROVED** [review]

**Comment by deeplow** [comment]
Waiting on OpenQA results before merging.

**Comment by legoktm** [comment]
We are green!
